pkg/download/download.go

```diff
@@ -0,0 +1,243 @@
+package download
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/firmwareci/contest-client/pkg/env"
+	"github.com/jlaffaye/ftp"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// DownloadBinary parses the unparsed URL, and proceeds to download
+func DownloadBinary(unparsedURL string) (string, error) {
+	URL, err := url.Parse(unparsedURL)
+	if err != nil {
+		return "", err
+	}
+	filename := path.Base(URL.Path)
+
+	binaryDirectory := os.Getenv(env.EnvBinDir)
+
+	if binaryDirectory != "" {
+		if path.IsAbs(binaryDirectory) != true {
+			return "", fmt.Errorf("Path: '%s', is not absolute", binaryDirectory)
+		}
+	} else {
+		binaryDirectory = env.DefaultBinaryDir
+	}
+
+	binaryPath := path.Join(binaryDirectory, filename)
+
+	file, err := os.Create(binaryPath)
+	if err != nil {
+		return "", err
+	}
+
+	defer file.Close()
+
+	switch URL.Scheme {
+	case "http", "https", "":
+		{
```

Comment:
These are actually not needed.

Reply:
I think they make this whole section a bit more readable and structured. Do you want me to delete them?